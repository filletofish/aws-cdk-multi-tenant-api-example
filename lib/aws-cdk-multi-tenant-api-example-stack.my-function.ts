import { Context, APIGatewayProxyResult, APIGatewayEvent } from 'aws-lambda';

export const handler = async (event: APIGatewayEvent, context: Context): Promise<APIGatewayProxyResult> => {
  console.log(`Event: ${JSON.stringify(event, null, 2)}`);
  console.log(`Context: ${JSON.stringify(context, null, 2)}`);

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const callerArn = event.requestContext.identity.userArn!;
  console.log(`Caller AWS ARN: ${callerArn}`);

  // Randomly throwing an error to demo how errors are handled in metrics / access logs
  const randomNumber = Math.random();
  if (randomNumber < 0.5) {
    throw new Error('Random internal error occurred!');
  }

  const callerAwsAccount = extractAWSAccountFromARN(callerArn);
  return {
    statusCode: 200,
    body: JSON.stringify({
      message: `Received API Call from AWS account=${callerAwsAccount}`,
    }),
  };
};

export const extractAWSAccountFromARN = (arn: string): string | null => {
  const regex = /^arn:aws:[a-z-]+::(\d+):.*$/;
  const match = arn.match(regex);
  if (match && match.length === 2) {
    return match[1];
  }
  return null;
};
