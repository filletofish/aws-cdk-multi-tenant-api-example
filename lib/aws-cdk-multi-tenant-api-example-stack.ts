import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { LambdaRestApi } from 'aws-cdk-lib/aws-apigateway';
import * as lambda from 'aws-cdk-lib/aws-lambda-nodejs';

export class MultiTenantApiGatewayExampleStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const testFunction = new lambda.NodejsFunction(this, 'my-function');

    new LambdaRestApi(this, 'RestApi', {
      handler: testFunction,
      cloudWatchRole: true,
    });
  }
}
