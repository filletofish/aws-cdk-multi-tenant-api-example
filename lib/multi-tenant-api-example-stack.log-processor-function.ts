import { Context, CloudWatchLogsEvent, Callback, CloudWatchLogsDecodedData } from 'aws-lambda';
import * as zlib from 'zlib';

import { MetricResolution, Metrics, MetricUnits } from '@aws-lambda-powertools/metrics';

const metrics = new Metrics({
  namespace: 'MultiTenantApi',
  serviceName: 'multi-tenant-api',
});

interface ApiAccessLogEvent {
  requestId: string;
  path: string;
  httpMethod: string;
  userArn: string;
  status: string;
  responseLatency: string;
}

export const handler = (event: CloudWatchLogsEvent, context: Context, callback: Callback) => {
  console.log('Event: ' + JSON.stringify(event, null, 2));
  const decoded = Buffer.from(event.awslogs.data, 'base64');
  zlib.gunzip(decoded, (e, result) => {
    if (e) {
      callback(e, null);
    } else {
      const json: CloudWatchLogsDecodedData = JSON.parse(result.toString('ascii'));
      json.logEvents.forEach((event) => {
        const apiAccessLogEvent: ApiAccessLogEvent = JSON.parse(event.message);

        console.log(`Parsed api access log event:` + JSON.stringify(apiAccessLogEvent));
        publishCustomMetricForAccessLogEvent(apiAccessLogEvent);
      });
      callback(null);
    }
  });
};

function createSingleMetricWithDefaultDimensions(accessLogEvent: ApiAccessLogEvent): Metrics {
  const singleMetric = metrics.singleMetric();

  singleMetric.addDimension('Path', accessLogEvent.path);
  const clientName = extractClientNameFromARN(accessLogEvent.userArn);
  singleMetric.addDimension('ClientName', clientName);
  singleMetric.addDimension('Method', accessLogEvent.httpMethod);

  return singleMetric;
}

// TODO: Specify timestamp of published metrics. Looks like EMF that used under the hood in EMF doesn't support setting metric timestamp
// https://github.com/awslabs/aws-lambda-powertools-python/issues/166
function publishCustomMetricForAccessLogEvent(accessLogEvent: ApiAccessLogEvent) {
  const countMetric = createSingleMetricWithDefaultDimensions(accessLogEvent);
  countMetric.addMetric('Count', MetricUnits.Count, 1, MetricResolution.High);

  const latency = parseInt(accessLogEvent.responseLatency);
  const latencyMetric = createSingleMetricWithDefaultDimensions(accessLogEvent);
  latencyMetric.addMetric('Latency', MetricUnits.Milliseconds, latency, MetricResolution.High);

  const statusCode = parseInt(accessLogEvent.status);
  if (statusCode >= 400 && statusCode < 500) {
    const clientErrorMetric = createSingleMetricWithDefaultDimensions(accessLogEvent);
    clientErrorMetric.addMetric('4XXError', MetricUnits.Count, 1, MetricResolution.High);
  }

  if (statusCode >= 500) {
    const serviceErrorMetric = createSingleMetricWithDefaultDimensions(accessLogEvent);
    serviceErrorMetric.addMetric('5XXError', MetricUnits.Count, 1, MetricResolution.High);
  }
}

const CLIENT_ROLE_ARN_REGEX = /^arn:aws:sts::\d+:assumed-role\/MultiTenantApiClient-(.+)-role\/\w+$/;

function isValidARN(value: string): boolean {
  const regex = CLIENT_ROLE_ARN_REGEX;
  return regex.test(value);
}

function extractClientNameFromARN(value: string): string {
  if (!isValidARN) {
    return 'unknown';
  }
  const match = value.match(CLIENT_ROLE_ARN_REGEX);
  if (match && match.length === 2) {
    return match[1];
  }

  return 'unknown';
}
