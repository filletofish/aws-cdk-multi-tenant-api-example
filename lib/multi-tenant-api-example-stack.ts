import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as apigw from 'aws-cdk-lib/aws-apigateway';
import * as lambda from 'aws-cdk-lib/aws-lambda-nodejs';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as cloudwatch from 'aws-cdk-lib/aws-cloudwatch';
import * as logsd from 'aws-cdk-lib/aws-logs-destinations';

interface ApiClientConfig {
  name: string;
  awsAccounts: string[];
  rateLimit: number;
  burstLimit: number;
}

const apiClients: ApiClientConfig[] = [
  {
    name: 'payment-service',
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    awsAccounts: [process.env.CDK_DEFAULT_ACCOUNT!],
    rateLimit: 10,
    burstLimit: 2,
  },
  {
    name: 'order-service',
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    awsAccounts: [process.env.CDK_DEFAULT_ACCOUNT!],
    rateLimit: 1,
    burstLimit: 1,
  },
];

export class MultiTenantApiExampleStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const apiFunction = new lambda.NodejsFunction(this, 'api-function', {
      functionName: 'multi-tenant-api-function',
    });

    const accessLogsGroup = new logs.LogGroup(this, 'RestApiLogs', {
      logGroupName: 'multi-tenant-api-access-logs',
    });

    const api = new apigw.LambdaRestApi(this, 'RestApi', {
      handler: apiFunction,
      cloudWatchRole: true,
      proxy: false,
      deployOptions: {
        accessLogFormat: this.apiGatewayAccessLoggingFormat(),
        accessLogDestination: new apigw.LogGroupLogDestination(accessLogsGroup),
      },
      endpointExportName: 'MultiTenantApiEndpoint',
    });

    this.createPerClientCountMetricFilter(accessLogsGroup);
    this.createApiAccessLogProcessor(accessLogsGroup);

    const resource = api.root.addResource('items');
    const getItemsMethod = resource.addMethod('GET', api.root.defaultIntegration, {
      authorizationType: apigw.AuthorizationType.IAM,
      apiKeyRequired: true,
    });

    const allowInvokeApiPolicy = new iam.Policy(this, 'AllowInvokeApi', {
      statements: [
        new iam.PolicyStatement({
          actions: ['execute-api:Invoke'],
          effect: iam.Effect.ALLOW,
          resources: [getItemsMethod.methodArn],
        }),
      ],
    });

    apiClients.forEach((apiClient) => this.setupApiClient(allowInvokeApiPolicy, api, apiClient));
  }

  private setupApiClient(allowInvokeApiPolicy: iam.Policy, api: apigw.RestApi, apiClient: ApiClientConfig) {
    const clientAccountPrincipals = apiClient.awsAccounts.map((acc) => new iam.AccountPrincipal(acc));
    const role = new iam.Role(this, 'MultiTenantApiClient-' + apiClient.name + '-Role', {
      roleName: `MultiTenantApiClient-${apiClient.name}-role`,
      assumedBy: new iam.CompositePrincipal(...clientAccountPrincipals),
    });
    role.attachInlinePolicy(allowInvokeApiPolicy);

    const plan = api.addUsagePlan('MultiTenantApiUsagePlan-' + apiClient.name, {
      name: apiClient.name + '-plan',
      throttle: {
        rateLimit: apiClient.rateLimit,
        burstLimit: apiClient.burstLimit,
      },
    });

    plan.addApiStage({
      stage: api.deploymentStage,
    });

    const key = api.addApiKey('MultiTenantApiKey-' + apiClient.name, {
      apiKeyName: `key-${apiClient.name}`,
      value: `api-key-${apiClient.name}`,
    });
    plan.addApiKey(key);
  }

  private createPerClientCountMetricFilter(accessLogsGroup: logs.LogGroup) {
    const metricNamespace = 'MultiTenantApiMetricFilter';

    new logs.MetricFilter(this, 'MultiTenantApiCountMetricFilter', {
      logGroup: accessLogsGroup,
      filterPattern: logs.FilterPattern.exists('$.userArn'),
      metricNamespace: metricNamespace,
      metricName: 'Count',
      metricValue: '1',
      unit: cloudwatch.Unit.COUNT,
      dimensions: {
        client: '$.userArn',
        method: '$.httpMethod',
        path: '$.path',
      },
    });

    new logs.MetricFilter(this, 'MultiTenantApiLatencyMetricFilter', {
      logGroup: accessLogsGroup,
      filterPattern: logs.FilterPattern.exists('$.userArn'),
      metricNamespace: metricNamespace,
      metricName: 'Latency',
      metricValue: '$.responseLatency',
      unit: cloudwatch.Unit.MILLISECONDS,
      dimensions: {
        client: '$.userArn',
        method: '$.httpMethod',
        path: '$.path',
      },
    });

    new logs.MetricFilter(this, 'MultiTenantApi4XXErrorMetricFilter', {
      logGroup: accessLogsGroup,
      filterPattern: logs.FilterPattern.literal('{ ($.status = 4**) && ($.userArn = "*") }'),
      metricNamespace: metricNamespace,
      metricName: '4XXError',
      metricValue: '1',
      unit: cloudwatch.Unit.COUNT,
      dimensions: {
        client: '$.userArn',
        method: '$.httpMethod',
        path: '$.path',
      },
    });

    new logs.MetricFilter(this, 'MultiTenantApi5XXErrorMetricFilter', {
      logGroup: accessLogsGroup,
      filterPattern: logs.FilterPattern.literal('{ ($.status = 5**) && ($.userArn = "*") }'),
      metricNamespace: metricNamespace,
      metricName: '5XXError',
      metricValue: '1',
      unit: cloudwatch.Unit.COUNT,
      dimensions: {
        client: '$.userArn',
        method: '$.httpMethod',
        path: '$.path',
      },
    });
  }

  private createApiAccessLogProcessor(accessLogsGroup: logs.LogGroup) {
    const logProcessingFunction = new lambda.NodejsFunction(this, 'log-processor-function', {
      functionName: 'multi-tenant-api-log-processor-function',
    });
    new logs.SubscriptionFilter(this, 'MultiTenantApiLogSubscriptionFilter', {
      logGroup: accessLogsGroup,
      destination: new logsd.LambdaDestination(logProcessingFunction),
      filterPattern: logs.FilterPattern.allEvents(),
    });
  }

  private apiGatewayAccessLoggingFormat(): apigw.AccessLogFormat {
    // See https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html#context-variable-reference

    const formatObject = {
      requestId: '$context.requestId',
      extendedRequestId: '$context.extendedRequestId',
      apiId: '$context.apiId',
      resourceId: '$context.resourceId',
      domainName: '$context.domainName',
      stage: '$context.stage',
      path: '$context.path',
      resourcePath: '$context.resourcePath',
      httpMethod: '$context.httpMethod',
      protocol: '$context.protocol',
      accountId: '$context.identity.accountId',
      sourceIp: '$context.identity.sourceIp',
      user: '$context.identity.user',
      userAgent: '$context.identity.userAgent',
      userArn: '$context.identity.userArn',
      caller: '$context.identity.caller',
      cognitoIdentityId: '$context.identity.cognitoIdentityId',
      status: '$context.status',
      integration: {
        // The status code returned from an integration. For Lambda proxy integrations, this is the status code that your Lambda function code returns.
        status: '$context.integration.status',
        // For Lambda proxy integration, the status code returned from AWS Lambda, not from the backend Lambda function code.
        integrationStatus: '$context.integration.integrationStatus',
        // The error message returned from an integration
        // A string that contains an integration error message.
        error: '$context.integration.error',
        latency: '$context.integration.latency',
      },
      error: {
        responseType: '$context.error.responseType',
        message: '$context.error.message',
      },
      requestTime: '$context.requestTime',
      responseLength: '$context.responseLength',
      responseLatency: '$context.responseLatency',
    };
    const accessLogFormatString = JSON.stringify(formatObject);
    return apigw.AccessLogFormat.custom(accessLogFormatString);
  }
}
