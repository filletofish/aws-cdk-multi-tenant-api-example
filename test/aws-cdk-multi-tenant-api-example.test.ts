import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { MultiTenantApiGatewayExampleStack } from '../lib/aws-cdk-multi-tenant-api-example-stack';

test('Multi Tenant Api Example Stack Matches Snapshot', () => {
  const app = new cdk.App();

  const stack = new MultiTenantApiGatewayExampleStack(app, 'MyTestStack');

  const template = Template.fromStack(stack);

  expect(template.toJSON()).toMatchSnapshot();
});
