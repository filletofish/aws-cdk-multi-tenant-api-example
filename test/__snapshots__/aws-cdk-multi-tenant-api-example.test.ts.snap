// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multi Tenant Api Example Stack Matches Snapshot 1`] = `
{
  "Mappings": {
    "DefaultCrNodeVersionMap": {
      "af-south-1": {
        "value": "nodejs16.x",
      },
      "ap-east-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-2": {
        "value": "nodejs16.x",
      },
      "ap-northeast-3": {
        "value": "nodejs16.x",
      },
      "ap-south-1": {
        "value": "nodejs16.x",
      },
      "ap-south-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-1": {
        "value": "nodejs16.x",
      },
      "ap-southeast-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-3": {
        "value": "nodejs16.x",
      },
      "ca-central-1": {
        "value": "nodejs16.x",
      },
      "cn-north-1": {
        "value": "nodejs16.x",
      },
      "cn-northwest-1": {
        "value": "nodejs16.x",
      },
      "eu-central-1": {
        "value": "nodejs16.x",
      },
      "eu-central-2": {
        "value": "nodejs16.x",
      },
      "eu-north-1": {
        "value": "nodejs16.x",
      },
      "eu-south-1": {
        "value": "nodejs16.x",
      },
      "eu-south-2": {
        "value": "nodejs16.x",
      },
      "eu-west-1": {
        "value": "nodejs16.x",
      },
      "eu-west-2": {
        "value": "nodejs16.x",
      },
      "eu-west-3": {
        "value": "nodejs16.x",
      },
      "me-central-1": {
        "value": "nodejs16.x",
      },
      "me-south-1": {
        "value": "nodejs16.x",
      },
      "sa-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-2": {
        "value": "nodejs16.x",
      },
      "us-gov-east-1": {
        "value": "nodejs16.x",
      },
      "us-gov-west-1": {
        "value": "nodejs16.x",
      },
      "us-iso-east-1": {
        "value": "nodejs14.x",
      },
      "us-iso-west-1": {
        "value": "nodejs14.x",
      },
      "us-isob-east-1": {
        "value": "nodejs14.x",
      },
      "us-west-1": {
        "value": "nodejs16.x",
      },
      "us-west-2": {
        "value": "nodejs16.x",
      },
    },
  },
  "Outputs": {
    "RestApiEndpoint0551178A": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "RestApi0C43BF4B",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "RestApiDeploymentStageprod3855DE66",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AllowInvokeApi0932B888": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "execute-api:Invoke",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":execute-api:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    {
                      "Ref": "RestApi0C43BF4B",
                    },
                    "/",
                    {
                      "Ref": "RestApiDeploymentStageprod3855DE66",
                    },
                    "/GET/items",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AllowInvokeApi0932B888",
        "Roles": [
          {
            "Ref": "RestApiClientpaymentserviceRole52B31700",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "RestApi0C43BF4B": {
      "Properties": {
        "Name": "RestApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "RestApiAccount7C83CF5A": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "RestApi0C43BF4B",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "RestApiCloudWatchRoleE3ED6605",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "RestApiClientpaymentserviceRole52B31700": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::507426004734:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": "rest-api-client-payment-service-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "RestApiCloudWatchRoleE3ED6605": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "RestApiDeployment180EC503599bc867318d6e60be213ad921747dd1": {
      "DependsOn": [
        "RestApiitemsGET2AAECCB8",
        "RestApiitems0D5B6650",
      ],
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "RestApiDeploymentStageprod3855DE66": {
      "DependsOn": [
        "RestApiAccount7C83CF5A",
      ],
      "Properties": {
        "AccessLogSetting": {
          "DestinationArn": {
            "Fn::GetAtt": [
              "RestApiLogsF6F215CD",
              "Arn",
            ],
          },
          "Format": "{"requestId":"$context.requestId","extendedRequestId":"$context.extendedRequestId","apiId":"$context.apiId","resourceId":"$context.resourceId","domainName":"$context.domainName","stage":"$context.stage","path":"$context.path","resourcePath":"$context.resourcePath","httpMethod":"$context.httpMethod","protocol":"$context.protocol","accountId":"$context.identity.accountId","sourceIp":"$context.identity.sourceIp","user":"$context.identity.user","userAgent":"$context.identity.userAgent","userArn":"$context.identity.userArn","caller":"$context.identity.caller","cognitoIdentityId":"$context.identity.cognitoIdentityId","status":"$context.status","integration":{"status":"$context.integration.status","integrationStatus":"$context.integration.integrationStatus","error":"$context.integration.error","latency":"$context.integration.latency"},"error":{"responseType":"$context.error.responseType","message":"$context.error.message"},"requestTime":"$context.requestTime","responseLength":"$context.responseLength","responseLatency":"$context.responseLatency"}",
        },
        "DeploymentId": {
          "Ref": "RestApiDeployment180EC503599bc867318d6e60be213ad921747dd1",
        },
        "RestApiId": {
          "Ref": "RestApi0C43BF4B",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "RestApiLogsF6F215CD": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "example-rest-api-access-logs",
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "RestApiitems0D5B6650": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "RestApi0C43BF4B",
            "RootResourceId",
          ],
        },
        "PathPart": "items",
        "RestApiId": {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "RestApiitemsGET2AAECCB8": {
      "Properties": {
        "AuthorizationType": "AWS_IAM",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "myfunctionDF753486",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "RestApiitems0D5B6650",
        },
        "RestApiId": {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "RestApiitemsGETApiPermissionMyTestStackRestApi13257D94GETitemsC33A05E9": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "myfunctionDF753486",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "RestApi0C43BF4B",
              },
              "/",
              {
                "Ref": "RestApiDeploymentStageprod3855DE66",
              },
              "/GET/items",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiitemsGETApiPermissionTestMyTestStackRestApi13257D94GETitems92CE32FB": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "myfunctionDF753486",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "RestApi0C43BF4B",
              },
              "/test-invoke-stage/GET/items",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "myfunctionDF753486": {
      "DependsOn": [
        "myfunctionServiceRole41691664",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "440b64ebfc461325fbc66a997fbe3f70ad0c8f5a85e264763ceb22fce2f6fc8b.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "myfunctionServiceRole41691664",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "DefaultCrNodeVersionMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "myfunctionServiceRole41691664": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
